refs/files-backend.c

```diff
@@ -3402,6 +3403,85 @@ static int files_ref_store_remove_on_disk(struct ref_store *ref_store,
 	return ret;
 }
 
+/*
+ * For refs and reflogs, they share a unified interface when scanning
+ * the whole directory. This function is used as the callback for each
+ * regular file or symlink in the directory.
+ */
+typedef int (*files_fsck_refs_fn)(struct fsck_refs_options *o,
+				  const char *gitdir,
+				  const char *refs_check_dir,
+				  struct dir_iterator *iter);
+
+static int files_fsck_refs_dir(struct ref_store *ref_store,
+			       struct fsck_refs_options *o,
+			       const char *refs_check_dir,
+			       files_fsck_refs_fn *fsck_refs_fns)
+{
+	struct dir_iterator *iter;
+	struct strbuf sb = STRBUF_INIT;
+	int ret = 0;
+	int iter_status;
+	const char *gitdir = ref_store->gitdir;
+
+	strbuf_addf(&sb, "%s/%s", gitdir, refs_check_dir);
+
+	iter = dir_iterator_begin(sb.buf, 0);
+
+	/*
+	 * The current implementation does not care about the worktree, the
+	 * worktree may have no refs/heads or refs/tags directory. Simply
+	 * return 0 now.
```

Comment:
How do we know that `!iter` indicates that there was a worktree? I think this is somewhat roundabout and may also be caused by various other reasons, like real repository corruption. We should likely verify up front whether the directory belongs to a worktree or not.

Reply:
Actually, I made a mistake in this part. Have fixed this.